pkg/autoscaler/statforwarder/forwarder.go

```diff
@@ -0,0 +1,253 @@
+	leaseInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: reconciler.NamespaceFilterFunc(ns),
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    f.leaseUpdated,
+			UpdateFunc: controller.PassNew(f.leaseUpdated),
+			// TODO(yanweiguo): Set up DeleteFunc.
+		},
+	})
+
+	return &f
+}
+
+func (f *Forwarder) leaseUpdated(obj interface{}) {
+	l := obj.(*coordinationv1.Lease)
+	ns, n := l.Namespace, l.Name
+
+	if !f.bs.HasBucket(n) {
+		// Skip this Lease as it is not for Autoscaler.
+		return
+	}
+
+	if l.Spec.HolderIdentity == nil || *l.Spec.HolderIdentity == "" {
+		f.logger.Debugf("HolderIdentity not found for Lease %s/%s", ns, n)
+		return
+	}
+
+	identity := *l.Spec.HolderIdentity
+	if h, ok := f.getHolder(n); ok && h == identity {
+		// Already up-to-date.
+		return
+	}
+
+	f.setHolder(n, identity)
+
+	if identity != f.selfIP {
+		// Skip creating/updating Service and Endpoints as not the holder.
+		return
+	}
+
+	if err := f.createService(ns, n); err != nil {
+		f.logger.Errorf("Failed to create Service for Lease %s/%s: %v", ns, n, err)
+		return
+	}
+	f.logger.Infof("Created Service for Lease %s/%s", ns, n)
+
+	if err := f.createOrUpdateEndpoints(ns, n); err != nil {
+		f.logger.Errorf("Failed to create Endpoints for Lease %s/%s: %v", ns, n, err)
+		return
+	}
+	f.logger.Infof("Created Endpoints for Lease %s/%s", ns, n)
+}
+
+func (f *Forwarder) checkService(ns, n string) error {
+	return retry.OnError(retry.DefaultRetry, exceptNotFound, func() error {
+		_, err := f.serviceLister.Services(ns).Get(n)
+		return err
+	})
+}
+
+func (f *Forwarder) createService(ns, n string) error {
+	err := f.checkService(ns, n)
+	if err == nil {
+		// Already created.
+		return nil
+	}
+
+	if !apierrs.IsNotFound(err) {
+		return err
+	}
+
+	return wait.ExponentialBackoff(retry.DefaultRetry, func() (bool, error) {
+		_, err = f.kc.CoreV1().Services(ns).Create(&v1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      n,
+				Namespace: ns,
+			},
+			Spec: v1.ServiceSpec{
+				Ports: []v1.ServicePort{{
+					Name:       autoscalerPortName,
+					Port:       autoscalerPort,
+					TargetPort: intstr.FromInt(autoscalerPort),
+				}},
+			},
+		})
+		return err == nil, err
+	})
+}
+
+func (f *Forwarder) getEndpoints(ns, n string) (*v1.Endpoints, error) {
+	var e *v1.Endpoints
+	err := retry.OnError(retry.DefaultRetry, exceptNotFound, func() error {
+		localE, err := f.endpointsLister.Endpoints(ns).Get(n)
+		e = localE
+		return err
+	})
+	return e, err
+}
+
+func (f *Forwarder) createOrUpdateEndpoints(ns, n string) error {
+	wantSubsets := []v1.EndpointSubset{{
+		Addresses: []v1.EndpointAddress{{
+			IP: f.selfIP,
+		}},
+		Ports: []v1.EndpointPort{{
+			Name:     autoscalerPortName,
+			Port:     autoscalerPort,
+			Protocol: v1.ProtocolTCP,
+		}}},
+	}
+
+	e, err := f.getEndpoints(ns, n)
+	// Update existing Endpoints.
+	if err == nil {
+		return reconciler.RetryUpdateConflicts(func(attempts int) (err error) {
```

Comment:
Is it possible to have one exponential backoff instead of two? If I'm not missing something, can we get the endpoints here, and keep retrying if we are not getting a not found error?

Reply:
Good point. Yeah in updating retry loop we also call get so we can merge them. Done.